earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp

```diff
@@ -278,16 +278,105 @@ class StateUpdater::SetStateVisitor : public default_dfs_visitor {
+          }
+        }
+        else {
+          // This shoud never happen - we had to successfully load the asset
+          // previously to get it into the tree.
+          notify(NFY_WARN, "Could not load asset '%s' to set state.",
+                name.toString().c_str());
+        }
+      }
+    }
+
+  public:
+    SetBlockingStateVisitor(StateUpdater & updater, DependentStateTree & tree, AssetDefs::State newState) :
+        updater(updater), tree(tree), newState(newState),
+        hasBlockingInputs(std::make_shared<RecalcSet>()),
+        hasBlockingChildren(std::make_shared<RecalcSet>()) {}
+
+    // This function is called after the DFS has completed for every vertex
+    // below this one in the tree. Thus, we don't calculate the state for an
+    // asset until we've calculated the state for all of its inputs and
+    // children.
+    void finish_vertex(
+        DependentStateTreeVertexDescriptor vertex,
+        const DependentStateTree &) const {
+      const AssetVertex & data = tree[vertex];
+      notify(NFY_PROGRESS, "Calculating state for '%s'", data.name.toString().c_str());
+
+      // Set the state for assets in the dependent tree.
+      if (data.inDepTree) {
+        SetState(vertex, newState, {0, 0});
+      }
+      else {
+        // Set the state to blocked if needed, otherwise do nothing
+        auto version = updater.storageManager->Get(data.name);
+        if (hasBlockingInputs->find(data.name) != hasBlockingInputs->end() && 
+            version->InputStatesAffectMyState(AssetDefs::Blocked, true)) {
+          SetState(vertex, AssetDefs::Blocked, {0,0});
+        }
+        else if (hasBlockingChildren->find(data.name) != hasBlockingChildren->end() && 
+            version->ChildStatesAffectMyState()) {
+          SetState(vertex, AssetDefs::Blocked, {0,0});
+        }
+      }
+    }
+};
+
+/**********************************************
+* StateUpdater
+**********************************************/
 void StateUpdater::SetStateForRefAndDependents(
     const SharedString & ref,
     AssetDefs::State newState,
     function<bool(AssetDefs::State)> updateStatePredicate) {
   try {
     SharedString verref = AssetVersionImpl::Key(ref);
-    bool includeDepDescendents = !UserActionRequired(newState);
+    bool includeAllChildrenAndInputs = !UserActionRequired(newState);
+    // For all states except Failed, we will load dependent children into the tree
+    bool includeDependentChildren = newState != AssetDefs::Failed;
     DependentStateTree tree = BuildDependentStateTree(
-        verref, updateStatePredicate, includeDepDescendents, storageManager);
-    depth_first_search(tree, visitor(SetStateVisitor(*this, tree, newState)));
+        verref, updateStatePredicate, includeDependentChildren, includeAllChildrenAndInputs, storageManager);
+      
+    if (AssetDefs::Canceled == newState || AssetDefs::Failed == newState)
```

Comment:
Minor: I feel like there should be an "IsBlocking" function you can pass a state to. There isn't one, but it might come in handy in the future.

Reply:
There's `CanRebuild` which looks like the same thing:
```
static bool CanRebuild(State state) {
   return state & (AssetDefs::Canceled | AssetDefs::Failed);
 }
```
Could rename that or just add `IsBlocking` and have `CanRebuild` call that.

Reply:
Nevermind, I guess `IsBlocking` would be a bit different. `Blocked` is also a blocking state.